File: Source/GraphAlgorithms/is_dag.js
let graph = [];
let visited = [];
let n;

function isCyclic(i) {
    if (visited[i] === 2) {
        return true;
    }
    visited[i] = 1;
    for (let j = 0; j < graph[i].length; j++) {
        let node = graph[i][j];
        if (visited[node] === 1) return true;
        else if (visited[node] === 0 && isCyclic(node)) return true;
    }
    visited[i] = 2;
    return false;
}

function isDAG() {
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            visited[j] = 0;
        }
        if (isCyclic(i)) {
            return false;
        }
    }
    return true;
}

function main() {
    n = 4;
    graph = [[1], [2], [3], []];
    console.log("Is DAG: " + (isDAG() ? "Yes" : "No"));
}

main();

File: Source/GraphAlgorithms/is_dag.c
#include <stdio.h>
#define MAX 100

int graph[MAX][MAX];
int visited[MAX];
int n;

int isCyclic(int i) {
    if (visited[i] == 2) {
        return 1;
    }
    visited[i] = 1;
    for (int j = 0; j < n; j++) {
        if (graph[i][j]) {
            if (visited[j] == 1) return 1;
            else if (visited[j] == 0 && isCyclic(j)) return 1;
        }
    }
    visited[i] = 2;
    return 0;
}

int isDAG() {
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            visited[j] = 0;
        }
        if (isCyclic(i)) {
            return 0;
        }
    }
    return 1;
}
